keras/layers/ntm.py

```diff
@@ -0,0 +1,320 @@
+import numpy as np
+from scipy.linalg import circulant
+
+import theano
+import theano.tensor as T
+floatX = theano.config.floatX
+
+from keras.layers.recurrent import Recurrent, GRU, LSTM
+from keras.utils.theano_utils import shared_zeros, alloc_zeros_matrix, shared_scalar
+tol = 1e-4
+
+
+def _update_controller(self, inp, h_tm1, M, mask):
+    """We have to update the inner RNN inside the NTM, this
+    is the function to do it. Pretty much copy+pasta from Keras
+    """
+    x = T.concatenate([inp, M], axis=-1)
+    # get inputs
+    if self.inner_rnn == 'gru':
+        x_z = T.dot(x, self.rnn.W_z) + self.rnn.b_z
+        x_r = T.dot(x, self.rnn.W_r) + self.rnn.b_r
+        x_h = T.dot(x, self.rnn.W_h) + self.rnn.b_h
+
+    elif self.inner_rnn == 'lstm':
+        xi = T.dot(x, self.rnn.W_i) + self.rnn.b_i
+        xf = T.dot(x, self.rnn.W_f) + self.rnn.b_f
+        xc = T.dot(x, self.rnn.W_c) + self.rnn.b_c
+        xo = T.dot(x, self.rnn.W_o) + self.rnn.b_o
+
+    elif self.inner_rnn == 'simple':
+        x = T.dot(x, self.rnn.W) + self.rnn.b
+
+    # update state
+    if self.inner_rnn == 'gru':
+        h = self.rnn._step(x_z, x_r, x_h, 1., h_tm1[0],
+                           self.rnn.U_z,
+                           self.rnn.U_r,
+                           self.rnn.U_h)
+        h = mask[:, None] * h + (1-mask[:, None])*h_tm1[0]
+        h = (h, )
+
+    elif self.inner_rnn == 'lstm':
+        h = self.rnn._step(xi, xf, xo, xc, 1.,
+                           h_tm1[1], h_tm1[0],
+                           self.rnn.U_i, self.rnn.U_f,
+                           self.rnn.U_o, self.rnn.U_c)
+        h = h[::-1]
+        h = tuple([mask[:, None]*h[i] +
+                   (1-mask[:, None])*h_tm1[i] for i in range(len(h))])
+
+    elif self.inner_rnn == 'simple':
+        h = self.rnn._step(x, 1, h_tm1[0], self.rnn.U)
+        h = mask[:, None] * h + (1-mask[:, None])*h_tm1[0]
+        h = (h, )
+
+    return h
+
+
+def _circulant(leng, n_shifts):
+    """
+    I confess, I'm actually proud of this hack. I hope you enjoy!
```

Comment:
In the near future, I hope we can have full in-class docstrings throughout the codebase and have an autogenerated documentation website. That will mean an obligation of sobriety and professionalism in the comments ;-)